pkg/gpu/cgroups.go

```diff
@@ -8,97 +8,344 @@
+		return "", fmt.Errorf("failed to get cgroups for pid %d: %w", pid, err)
+	}
 
-const (
-	systemdDev deviceType = "systemd"
-	cgroupDev  deviceType = "cgroup"
-)
+	if len(procCgroups) == 0 {
+		return "", fmt.Errorf("no cgroups found for pid %d", pid)
+	}
+
+	// Each cgroup is for a different subsystem in cgroupv1, we only want the cgroup ID
+	// and we can extract that from any cgroup
+	// In cgroupv2 we only have one cgroup, so this code also works.
+	cgroupPath := string(procCgroups[0].Path)
+
+	// If we're running in the host (no path to the host root filesystem), we can
+	// just return the cgroup path as we see it, we cannot do anything else.
+	// Also, cgroupv1 returns the cgroup name correctly here, so we can return it
+	// directly too
+	if rootfs == "" || cgroups.Mode() == cgroups.Legacy {
+		return cgroupPath, nil
+	}
+
+	// Now we need to deal with possibly different cgroup namespaces, which can
+	// happen in containerized environments. The cgroups we see from the
+	// system-probe container are different from the cgroups we see from the
+	// host. The cgroup extracted in the code above from /proc/pid/cgroup will
+	// be valid only for the container cgroup namespace, but we want the cgroup
+	// name in the host cgroup namespace.
+	//
+	// There are several possibilities to achieve this. The first, and easiest
+	// to implement, would be to enter the root cgroup namespace and look at the
+	// cgroup then. Unfortunately, this is not always possible as GKE seems to
+	// deny containers the access to the root cgroup namespace even with
+	// CAP_SYS_ADMIN capabilities.
+	//
+	// The other option is to match the container cgroup to a cgroup in the host
+	// namespace using the cgroup directory from the host, the one that can be
+	// seen in rootfs. Despite cgroups not having canonical IDs/names in their
+	// directories, the inodes are unique and constant among namespaces.
+	//
+	// However, this only works as long the cgroup name returned in
+	// /proc/pid/cgroup is absolute (i.e, /something). If it's relative (i.e,
+	// /../something, which can happen if the target process is in a cgroup that
+	// is a sibling of the current process' cgroup), then we need to resolve
+	// first our own cgroup path and resolve the relative path based on that.
+	pathParts := strings.Split(cgroupPath, "/")
+
+	if len(pathParts) > 1 && pathParts[1] == ".." { // first part is an empty string as the cgroup path starts by /
+		// Sanity check that we're not recursively getting our own cgroup, avoiding infinite recursion
+		currentPid := uint32(os.Getpid())
+		if pid == currentPid {
+			return "", fmt.Errorf("impossible situation: got a relative path for our own cgroup %s for pid %d", cgroupPath, pid)
+		}
+
+		currentCgroup, err := getAbsoluteCgroupForProcess(rootfs, currentPid)
+		if err != nil {
+			return "", fmt.Errorf("failed to get current process (pid=%d) cgroup: %w", currentPid, err)
+		}
+
+		return filepath.Join(currentCgroup, cgroupPath), nil
+	}
```

Comment:
should we sanitize the path to avoid path traversal, (`cgroupPath = "../../../etc/passwd"`)